### PR TITLE
🎨 Introduce ordered collections and simplify `.mapped()`

### PR DESCRIPTION
**Pull request by falexwolf**
- https://github.com/laminlabs/lnschema-core/pull/333

### CODE REVIEW

File: tests/test_collection.py [hunk 1]
```diff
@@ -314,8 +312,8 @@ def test_collection_mapped():
     assert len(ls_ds[0]) == 2 and len(ls_ds[2]) == 2
 
     # adata3 goes first here
-    with collection_outer.mapped(label_keys="feat1", join_vars="auto") as ls_ds:
-        assert ls_ds.join_vars == "outer"
+    with collection_outer.mapped(label_keys="feat1", join="inner") as ls_ds:
+        assert ls_ds.join_vars == "inner"
```
Comment (Koncopd) [inline tests/test_collection.py:316]: collection_outer has adatas with non-intersecting vars, this collection should be used to test outer join, not inner.

**Comment by Koncopd** [comment]
I don't think removing "auto" is necessary, it is easier to check with "auto" first then to manually check variables. Also it is important to avoid any join if the variables are exactly the same for performance reasons.

**Comment by Koncopd** [comment]
I believe reoving join_vars="auto" doesn't simplify anything, just requires a user to do additional actions manually instead.

File: tests/test_collection.py (hunk 1 above)
Comment (falexwolf) [reply tests/test_collection.py]: Fixed!

**Comment by falexwolf** [comment]
Pasting our discussion for reference: https://laminlabs.slack.com/archives/C03Q5TXF797/p1704653536322439?thread_ts=1704228794.520039&cid=C03Q5TXF797

File: tests/test_collection.py [hunk 2]
```diff
@@ -314,18 +317,29 @@ def test_collection_mapped():
...
 
     # adata3 goes first here
-    with collection_outer.mapped(label_keys="feat1", join_vars="auto") as ls_ds:
+    with collection_outer.mapped(label_keys="feat1", join="outer") as ls_ds:
         assert ls_ds.join_vars == "outer"
         assert len(ls_ds.var_joint) == 6
         assert len(ls_ds[0]) == 2
         assert len(ls_ds[0][0]) == 6
-        assert np.array_equal(ls_ds[1][0], np.array([4, 5, 8, 0, 0, 0]))
+        print(ls_ds[0][0])
```
Comment (Zethson) [inline tests/test_collection.py:325]: Are those print statements a relict from making the tests work?

Comment (falexwolf) [reply tests/test_collection.py:325]: Yes, and a slight worry that this output might be helpful in the future again. It took me a while to understand the test, and so I thought I'd leave these statements there for a little longer.

**Pull request merged**